source/modest/finder/pseudo_class.c

```diff
@@ -77,6 +77,62 @@ bool modest_finder_selector_sub_type_pseudo_class_function_has(modest_finder_t*
     return false;
 }
 
+// concat str1 and str2
+char *concat_string(const char *str1, const char *str2)
+{
+    char *result = NULL;
+    size_t n = 0;
+
+    if(str1) n += strlen(str1);
+    if(str2) n += strlen(str2);
+
+    if((str1 || str2) && (result = malloc(n + 1)) != NULL)
+    {
+        *result = '\0';
+
+        if(str1) strcpy(result, str1);
+        if(str2) strcat(result, str2);
+    }
+
+    return result;
+}
+
+bool modest_finder_selector_sub_type_pseudo_class_function_contains(modest_finder_t* finder, myhtml_tree_node_t* base_node, mycss_selectors_entry_t* selector, mycss_selectors_specificity_t* spec)
+{
+    if(base_node){
+        const char* text = NULL;
+        myhtml_tree_node_t *text_node = myhtml_node_child(base_node);
+        if(text_node) {
+          text = myhtml_node_text(text_node, NULL);
+          if(text){
+            char *data = NULL;
+            mycss_selectors_list_t *list = selector->value;
+            for(size_t i = 0; i < list->entries_list_length; i++) {
+                bool i_found = false;
+                mycss_selectors_entry_t *sel_entry = list->entries_list[i].entry;
+
+                if(sel_entry->key->data){
+                    data = concat_string(data, sel_entry->key->data);
+                }
+                
+                mycss_selectors_entry_t *next = sel_entry->next;
+                while(next && !i_found){
+                    if(next->key->data){
+                        data = concat_string(data, " ");
```

Comment:
memory leak!
if we allocation memory in `concat_string` and call `concat_string` with returned data we get leak.

Reply:
Of course :)